types/jsforce/salesforce-object.d.ts

```diff
@@ -3,18 +3,20 @@ import * as stream from 'stream';
 import { SObjectCreateOptions } from './create-options';
 import { DescribeSObjectResult } from './describe-result';
 import { Query } from './query';
-import { Record } from './record';
+import { Record, RecordReference } from './record';
 import { RecordResult } from './record-result';
 import { Connection } from './connection';
 import { SalesforceId } from './salesforce-id';
 
-export class SObject {
-    record(options: any, callback?: (err: Error, ret: any) => void): void;
-    update(options: SObjectCreateOptions, callback?: (err: Error, ret: any) => void): void;
-    retrieve(ids: string | string[], callback?: (err: Error, ret: Record | Record[]) => void): Promise<Record | Record[]>;
-    retrieve(ids: string | string[], options?: Object, callback?: (err: Error, ret: Record | Record[]) => void): Promise<Record | Record[]>;
-    upsert(records: Record | Record[], extIdField: SalesforceId, options?: Object, callback?: (err: Error, ret: RecordResult) => void): Promise<RecordResult | RecordResult[]>;
-    upsertBulk(input?: Record[] | stream.Stream | string, callback?: (err: Error, ret: RecordResult) => void): Batch;
+export class SObject<T> {
+    record(id: SalesforceId): RecordReference<T>;
+    retrieve(id: SalesforceId, options?: Object, callback?: (err: Error, record: Record<T>) => void): Promise<Record<T>>;
+    retrieve(ids: SalesforceId[], options?: Object, callback?: (err: Error, ret: Array<Record<T>>) => void): Promise<Array<Record<T>>>;
+    update(record: Partial<T>, options?: Object, callback?: (err: Error, ret: RecordResult) => void): Promise<RecordResult>;
```

Comment:
I split out single and array versions of these methods so that the return types could be more strictly provided.